variantlib/commands/make_variant.py

```diff
@@ -0,0 +1,222 @@
+            with open(wheel_file_path, "wb") as f:  # noqa: PTH123
+                whl_pck.BytesGenerator(f, maxheaderlen=0).flatten(info)
+
+    # Reassemble the tags for the wheel file
+    tagline = whl_pck.compute_tagline(tags)
+
+    # Repack the wheel
+    wheel_path = os.path.join(dest_dir, f"{name_version}-{tagline}-{variant_hash}.whl")  # noqa: PTH118
+    with whl_pck.WheelFile(wheel_path, "w") as wf:
+        logging.info(
+            "Repacking wheel as `%(wheel_path)s` ...", {"wheel_path": wheel_path}
+        )
+        wf.write_files(directory)
+
+    return wheel_path
+
+
+def make_variant(args: list[str]) -> None:
+    parser = argparse.ArgumentParser(
+        prog="make_variant",
+        description="Transform a normal Wheel into a Wheel Variant.",
+    )
+
+    parser.add_argument(
+        "-p",
+        "--property",
+        dest="properties",
+        action="extend",
+        nargs="+",
+        help=(
+            "Variant Properties to add to the Wheel Variant, can be repeated as many "
+            "times as needed"
+        ),
+    )
+
+    parser.add_argument(
+        "-f",
+        "--file",
+        dest="input_filepath",
+        type=pathlib.Path,
+        required=True,
+        help="Wheel file to process",
+    )
+
+    parser.add_argument(
+        "-o",
+        "--output_directory",
+        type=pathlib.Path,
+        required=True,
+        help="Output Directory to use to store the Wheel Variant",
+    )
+
+    parsed_args = parser.parse_args(args)
+
+    wheel_file_re = re.compile(
+        r"""^(?P<namever>(?P<name>[^\s-]+?)-(?P<ver>[^\s-]*?))
+        ((-(?P<build>\d[^-]*?))?-(?P<pyver>[^\s-]+?)-(?P<abi>[^\s-]+?)-(?P<plat>[^\s-]+?)
+        \.whl|\.dist-info)$""",
+        re.VERBOSE,
+    )
+
+    input_filepath: pathlib.Path = parsed_args.input_filepath
+    output_directory: pathlib.Path = parsed_args.output_directory
+
+    # Input Validation
+    if not input_filepath.is_file():
+        raise FileExistsError(f"Input Wheel File `{input_filepath}` does not exists.")
+
+    if not output_directory.is_dir():
+        raise FileExistsError(f"Output Directory `{output_directory}` does not exists.")
+
+    # Input Validation - Wheel Filename is valid and non variant already.
+    wheel_info = wheel_file_re.match(input_filepath.name)
+    if not wheel_info:
+        raise ValueError(f"`{input_filepath.name}` is not a valid wheel filename.")
+
+    # Transform properties into a VariantDescription
+    properties_args: list[str] | None = parsed_args.properties
+    if properties_args is None or len(properties_args) == 0:
+        raise ValueError("At least one Variant Property is required.")
+
+    vprops = [VariantProperty.from_str(vprop) for vprop in properties_args]
+    vdesc = VariantDescription(properties=vprops)
+
+    with tempfile.TemporaryDirectory() as _tmpdir:
+        tempdir = pathlib.Path(_tmpdir)
+        wheel_unpack(input_filepath, tempdir)
```

Comment:
I don't think we really need to unpack it. Unfortunately, it looks like `zipfile` lacks proper API to update files in place or copy compressed members (https://github.com/python/cpython/pull/125718), but at least we could avoid writing files to disk and potentially altering their metadata in the process.

I.e. basically open both archives, iterate over members: if not metadata, copy between archives as-is; if metadata, read, alter and write. It will probably even be simpler than the current logic, which seems to make unnecessary changes to the archive (probably because it was adapted from the command altering tags, which we don't do here).

Reply:
I don't disagree though it's a little complicated for a first iteration - feel free to experiment and optimize ;)